### PR TITLE
[BEAM-8086] add isRetraciton() to WindowedValue.

### PR DESCRIPTION
**Pull request by amaliujia**
We need a way to mark WindowedValue as a retraction. Current idea is we should separate sub-classes of WindowedValue without and with retraction flag to reduce the impact of enabling retractions. It is mostly because I plan to use Boolean as the flag of retraction so each instance of WindowedValue will have 8 bytes overhead if we don't have such a separation.

In the future, as the implementation of retraction work goes, I will probably add more sub-classes of WindowedValue which implements isRetraction(), and those will be used for retracting mode. Hopefully by doing so, we can limit the overhead of message metadata change on pipelines that enables retraction only. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by amaliujia** [comment]
An open question is if the default return value should be `false` than throwing an exception. It could be left to the future for clarification.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/WindowedValue.java [hunk 1]
```diff
@@ -141,6 +141,11 @@ public boolean isSingleWindowedValue() {
     return false;
   }
 
+  /** Returns {@code true} if this WindowedValue is a retraction. */
+  public boolean isRetraction() {
```
Comment (amaliujia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/WindowedValue.java]: Should we add `@Expremental`? How to decide what piece of SDK is user facing?

**Comment by lukecwik** [comment]
LATE/EARLY/... status is part of PaneInfo, would that be a better place?

**Comment by amaliujia** [comment]
@lukecwik Thanks for the idea. It reminds me that I actually got this suggestion in #9199 as well.

Let me compare PaneInfo with WindowedValue and I will summarize pros and cons for the further discussion.

**Comment by je-ik** [comment]
I think, that `PaneInfo` and `RetractionMode` (let's say) are orthogonal. It is imaginable (at least in theory), that at fully general level there will be early retraction firings. That can happen at least with early firings on merging windows, because every merge with previously (early) fired data will have to be retracted.
I'd also suggest to convert the `boolean isRetraction()` to something returning and `enum`, because it is better extensible and is somehow more consistent with `PaneInfo`.

**Pull request closed without merging**